feat(api): improve isValidSpanId, isValidTraceId performance

seemk opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/guides/contributor#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Short description of the changes

Use a lookup table to check for hex string validity instead of regexes.

Changed a test where a string was cast into `SpanContext`. Previously the test called `isSpanContextValid` with a string instead of  `SpanContext`, so  `undefined` `spanId` and `traceId` were passed to `isValidTraceId` and `isValidSpanId`. The regex implementation handled this implicitly, but this had never been exposed to the types. If this is an issue I can add a check for an `undefined` input (even though according to the types the input can only be string).


Tested in two different ways:

Test 1 - Total time to process 50 runs of 1M pregenerated IDs
Test 2 - Output of the `benchmark` library

**Mixed IDs (equal amount of valid and invalid IDs)**

```
Test 1
isValidTraceId (original) 4810.5 ms
isValidTraceId (new)      1808.5 ms (2.66x)

isValidSpanId (original) 2994.1 ms
isValidSpanId (new)       920.1 ms (3.25x)

Test 2
isValidTraceId (original) x 10,292,556 ops/sec ±0.41% (98 runs sampled)
isValidTraceId (new)      x 24,541,032 ops/sec ±0.51% (97 runs sampled) (2.38x)

isValidSpanId (original) x 15,623,278 ops/sec ±0.48% (96 runs sampled)
isValidSpanId (new)      x 43,478,105 ops/sec ±0.87% (94 runs sampled) (2.78x)
```

**All valid IDs**

```
Test 1
isValidTraceId (original) 5216.2 ms
isValidTraceId (new)      1467.4 ms (3.55x)

isValidSpanId (original) 3026.1 ms
isValidSpanId (new)       760.1 ms (3.98x)

Test 2
isValidTraceId (original) x 9,073,249 ops/sec ±0.36% (96 runs sampled)
isValidTraceId (new)      x 29,349,911 ops/sec ±0.51% (97 runs sampled) (3.23x)

isValidSpanId (original) x 14,880,091 ops/sec ±0.59% (95 runs sampled)
isValidSpanId (new)      x 51,324,066 ops/sec ±0.78% (96 runs sampled) (3.45x)
```


**All random invalid IDs (randomly generated ID with an invalid character appearing at a random location)**

```
Test 1
isValidTraceId (original) 4254.9 ms
isValidTraceId (new)      2083.2 ms (2.04x)

isValidSpanId (original) 2518.9 ms
isValidSpanId (new)      1058.5 ms (2.38x)

Test 2
isValidTraceId (original) x 11,643,782 ops/sec ±0.44% (99 runs sampled)
isValidTraceId (new)      x 21,810,864 ops/sec ±0.71% (98 runs sampled) (1.87x)

isValidSpanId (original) x 18,313,326 ops/sec ±0.45% (96 runs sampled)
isValidSpanId (new)      x 39,067,315 ops/sec ±0.87% (95 runs sampled) (2.13x)
```


**INVALID_{TRACE,SPAN}_ID - all zeroes**

```
Test 1
isValidTraceId (original): 1540.1 ms
isValidTraceId (new):      1410.1 ms (1.09x)

isValidSpanId (original): 958.3 ms
isValidSpanId (new):      723.8 ms (1.32x)

Test 2
isValidTraceId (original) x 28,751,309 ops/sec ±0.78% (94 runs sampled)
isValidTraceId (new)      x 29,106,073 ops/sec ±0.71% (93 runs sampled) (1.01x)

isValidSpanId (original) x 41,279,268 ops/sec ±0.84% (96 runs sampled)
isValidSpanId (new)      x 50,268,626 ops/sec ±1.04% (92 runs sampled) (1.22x)
```

pichlermarc commented on api/src/trace/spancontext-utils.ts:
```diff
@@ -18,21 +18,41 @@ import { NonRecordingSpan } from './NonRecordingSpan';
 
-const VALID_TRACEID_REGEX = /^([0-9a-f]{32})$/i;
-const VALID_SPANID_REGEX = /^[0-9a-f]{16}$/i;
+const isHex = [
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
+  1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+] as const;
```
I think it can be helpful to add a comment here that this is a lookup table for ASCII characters 0-127, stating that valid hex chars are marked as `1`, non-valid hex chars are `0`.

seemk replied:
Added a comment

pichlermarc commented on api/src/trace/spancontext-utils.ts:
```diff
@@ -18,21 +18,41 @@ import { NonRecordingSpan } from './NonRecordingSpan';
+  let r = 0;
+  for (let i = 0; i < id.length; i += 4) {
+    r +=
+      isHex[id.charCodeAt(i) & 0x7f] +
```
maybe also a comment about masking so that this only lowest 7 bits are retained.

dyladan replied:
It is possible using `charCodeAt` to get partial code points from an extended-unicode code point containing string. Should we be using `codePointAt`? It would then also mean the number might be larger than 128 though (up to `2^16` I think)

seemk replied:
This comment got me thinking that this implementation had an error, where an UTF-16 value like `\u4141` (䅁) would be truncated to `0x41`, or `A`, which is a valid character and an invalid ID containing such value would pass.

I changed the implementation to `LUT[id.charCodeAt(i)] | 0` which [coerces undefined to 0](https://tc39.es/ecma262/#sec-numberbitwiseop) and fixes the issue (while being faster than the original). See the chart - `LUT unroll x 4` vs `LUT unroll x 4` old.

As for `codePointAt`, I can't think of a case where a Unicode character taking up multiple UTF-16 values would have all of the UTF-16 characters in the range in the lookup table ('a-z', 'A-Z, '0-9'). Besides that, it seems [`String.length` should not be used with `codePointAt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt#looping_with_codepointat).

I've profiled different implementations now, here's the data:
![image](https://github.com/user-attachments/assets/252e60c1-c62a-4d90-a844-ecea7f724829)
![image](https://github.com/user-attachments/assets/292fd309-ef35-4a2f-b297-97f4be479d79)

The implementation in the PR is `U8 LUT, unroll x 4`.

You can try it out yourself: https://gist.github.com/seemk/f00c432d8779a03dbe1a28236f9dd198

pichlermarc reviewed (commented):
Nice!

I think some information about that look-up table could be helpful for someone that comes across this code wondering what it is (I happened to have seen this elsewhere but I don't know if it's common to see this sort of thing in JS).

> If this is an issue I can add a check for an undefined input (even though according to the types the input can only be string).

My initial impulse would be to leave it out and see if it's a problem, but let's be safe and add a check since it's the API package. 😅

seemk commented:
> Nice!
> 
> I think some information about that look-up table could be helpful for someone that comes across this code wondering what it is (I happened to have seen this elsewhere but I don't know if it's common to see this sort of thing in JS).
> 
> > If this is an issue I can add a check for an undefined input (even though according to the types the input can only be string).
> 
> My initial impulse would be to leave it out and see if it's a problem, but let's be safe and add a check since it's the API package. 😅

I reverted the test and added a check for undefined (plus the comments)

dyladan reviewed (commented):
Would it be just as fast and easier to read if it was a set?

```typescript
const hexChars = new Set([
  "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
  "a", "b", "c", "d", "e", "f",
  "A", "B", "C", "D", "E", "F",
]);

function isValidHex(id: string, length: number) {
  // As of 1.9.0 the id was allowed to be undefined,
  // even though it was not possible in the types.
  if (id === undefined || id.length !== length) return false;
  for (const c of id) {
    if (!hexChars.has(c)) {
      return false;
    }
  }
  return true;
}
```

seemk commented:
> Would it be just as fast and easier to read if it was a set?
> 
> ```ts
> const hexChars = new Set([
>   "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
>   "a", "b", "c", "d", "e", "f",
>   "A", "B", "C", "D", "E", "F",
> ]);
> 
> function isValidHex(id: string, length: number) {
>   // As of 1.9.0 the id was allowed to be undefined,
>   // even though it was not possible in the types.
>   if (id === undefined || id.length !== length) return false;
>   for (const c of id) {
>     if (!hexChars.has(c)) {
>       return false;
>     }
>   }
>   return true;
> }
> ```

It's actually slower than the regex version:

```
isValidTraceId (original) 5289.4 ms
isValidTraceId (LUT, unroll x4) 1471.3 ms
isValidTraceId (set) 15970.2 ms

isValidSpanId (original) 2917.4 ms
isValidSpanId (LUT, unroll x4) 751.4 ms
isValidSpanId (set) 8467.6 ms
---
isValidTraceId (original) x 9,136,152 ops/sec ±0.46% (98 runs sampled)
isValidTraceId (LUT, unroll x4) x 29,705,793 ops/sec ±0.69% (97 runs sampled)
isValidTraceId (set) x 3,032,105 ops/sec ±0.10% (100 runs sampled)

isValidSpanId (original) x 15,580,277 ops/sec ±0.50% (95 runs sampled)
isValidSpanId (LUT, unroll x4) x 53,550,772 ops/sec ±0.76% (93 runs sampled)
isValidSpanId (set) x 5,930,540 ops/sec ±0.19% (98 runs sampled)
```

seemk commented on api/src/trace/spancontext-utils.ts:
```diff
@@ -18,21 +18,45 @@ import { NonRecordingSpan } from './NonRecordingSpan';
+function isValidHex(id: string, length: number): boolean {
+  // As of 1.9.0 the id was allowed to be undefined,
+  // even though it was not possible in the types.
+  if (id === undefined || id.length !== length) return false;
```
Good point, changed it to `typeof id !== 'string'`

dyladan reviewed (approved):
I'm marking this approved but I have a couple more options that I think would be at least interesting to see the benchmark results since you already have a benchmark setup. You don't need to take either of them, I just think this type of micro-optimization is fun and you nerd-sniped me a bit.

The first version uses simple arithmetic instead of a LUT which may be more readable.

```js
function isHexArithmetic(str) {
  for (let i = 0; i < str.length; i++) {
    const charCode = str.charCodeAt(i);
    if (charCode < 48 || (charCode > 57 && charCode < 65) || (charCode > 70 && charCode < 97) || charCode > 102) {
      return false;
    }
  }
  return true;
}
```

The second version takes advantage of the contiguous memory of TypedArrays and completely throws readability out the window

```js
const LUT = new Uint8Array([
  0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000,
  0b11111111, 0b00000011, 0b01111110, 0b00000000, 0b00000000, 0b00000000,
  0b01111110, 0b00000000, 0b00000000, 0b00000000,
]);

function isHexPackedLUT(str) {
  for (let i = 0; i < str.length; i++) {
    const charCode = str.codePointAt(i);
    if (!(LUT[Math.floor(charCode >> 3)] & (1 << (charCode & 7)))) {
      return false;
    }
  }
  return true;
}
```

seemk commented:
> I'm marking this approved but I have a couple more options that I think would be at least interesting to see the benchmark results since you already have a benchmark setup. You don't need to take either of them, I just think this type of micro-optimization is fun and you nerd-sniped me a bit.
> 
> The first version uses simple arithmetic instead of a LUT which may be more readable.
> 
> ```js
> function isHexArithmetic(str) {
>   for (let i = 0; i < str.length; i++) {
>     const charCode = str.charCodeAt(i);
>     if (charCode < 48 || (charCode > 57 && charCode < 65) || (charCode > 70 && charCode < 97) || charCode > 102) {
>       return false;
>     }
>   }
>   return true;
> }
> ```
> 
> The second version takes advantage of the contiguous memory of TypedArrays and completely throws readability out the window
> 
> ```js
> const LUT = new Uint8Array([
>   0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000,
>   0b11111111, 0b00000011, 0b01111110, 0b00000000, 0b00000000, 0b00000000,
>   0b01111110, 0b00000000, 0b00000000, 0b00000000,
> ]);
> 
> function isHexPackedLUT(str) {
>   for (let i = 0; i < str.length; i++) {
>     const charCode = str.codePointAt(i);
>     if (!(LUT[charCode >> 3] & (1 << (charCode & 7)))) {
>       return false;
>     }
>   }
>   return true;
> }
> ```
> 
> edit: for those wondering what is going on in the packed LUT version
> 
> * `LUT` is a simple bit-array. It's just a tightly packed version of the one used in the PR. Notice that each byte is reverse order though. That's important because while the left-most byte is the first byte, the right-most bit is the first bit of each byte.
> * `>> 3` divides by 8 with no remainder to determine which byte we're looking at
> * `& 7` gets the remainder of division by 8 to determine which bit we're looking at in that byte
> * `1` is left shifted that many times to allow us to do the bitwise `&`

I tested an if else based one initially, but left it out, here are the results:

```
Valid hex strings (which sould be the usual case for OTel)
isValidTraceId (original) x 9,023,702 ops/sec ±0.41% (96 runs sampled)
isValidTraceId (LUT, unroll x4) x 29,370,600 ops/sec ±0.41% (97 runs sampled)
isValidTraceId (ifelse) x 8,792,365 ops/sec ±0.55% (99 runs sampled)
isValidTraceId (packed LUT) x 24,088,125 ops/sec ±0.54% (95 runs sampled)
isValidSpanId (original) x 15,017,101 ops/sec ±0.57% (94 runs sampled)
isValidSpanId (LUT, unroll x4) x 51,371,239 ops/sec ±0.63% (96 runs sampled)
isValidSpanId (ifelse) x 16,227,776 ops/sec ±0.58% (98 runs sampled)
isValidSpanId (packed LUT) x 42,877,635 ops/sec ±0.78% (98 runs sampled)
```

With malformed hex strings (invalid hex character at random position):

```
isValidTraceId (original) x 11,129,327 ops/sec ±0.43% (95 runs sampled)
isValidTraceId (LUT, unroll x4) x 29,276,959 ops/sec ±0.45% (96 runs sampled)
isValidTraceId (ifelse) x 16,068,503 ops/sec ±0.51% (97 runs sampled)
isValidTraceId (packed LUT) x 33,366,037 ops/sec ±0.95% (96 runs sampled)
isValidSpanId (original) x 17,646,360 ops/sec ±0.31% (97 runs sampled)
isValidSpanId (LUT, unroll x4) x 37,972,807 ops/sec ±0.64% (93 runs sampled)
isValidSpanId (ifelse) x 24,483,385 ops/sec ±0.53% (96 runs sampled)
isValidSpanId (packed LUT) x 41,557,560 ops/sec ±0.72% (95 runs sampled)
```

The packed LUT seems to do surprisingly well and is about 10% faster in the negative cases (likely due to the early exit) and only about 15-20% slower in the most common path.

dyladan commented on api/src/trace/spancontext-utils.ts:
```diff
@@ -18,21 +18,45 @@ import { NonRecordingSpan } from './NonRecordingSpan';
-const VALID_TRACEID_REGEX = /^([0-9a-f]{32})$/i;
-const VALID_SPANID_REGEX = /^[0-9a-f]{16}$/i;
+// Valid characters (0-9, a-f, A-F) are marked as 1.
+const isHex = [
```
I benchmarked this with jsbench.me and changing this to be a `Uint8Array` sped it up by about 12% for me

seemk replied:
It seems to be slightly faster on x64, see the results here: https://github.com/open-telemetry/opentelemetry-js/pull/5714#discussion_r2124745037 (there's a link to the benchmark script if you want to run it yourself).

I changed the implementation to use `Uint8Array`.
